rllib/examples/rl_modules/classes/autoregressive_actions_rlm.py

```diff
@@ -1,203 +1,135 @@
-    prior distribution. Then, the posterior action is sampled from a posterior
-    distribution that depends on the prior action and the input data. The prior
-    and posterior distributions are implemented as MLPs.
-
-    The following components are implemented:
-    - ENCODER: An encoder that processes the observations from the environment.
-    - PI: A Policy head that outputs the actions, the log probabilities of the
-        actions, and the input to the action distribution. This head is composed
-        of two sub-heads:
-        - A prior head that outputs the logits for the prior action distribution.
-        - A posterior head that outputs the logits for the posterior action
-            distribution.
-    - VF: A value function head that outputs the value function.
-
-    Note, this RLModule is implemented for the `PPO` algorithm only. It is not
-    guaranteed to work with other algorithms.
+class AutoregressiveActionsRLM(TorchRLModule, ValueFunctionAPI):
+    """An RLModule that uses an autoregressive action distribution.
+
+    Actions are sampled in two steps. The first (prior) action component is sampled from
+    a categorical distribution. Then, the second (posterior) action component is sampled
+    from a posterior distribution that depends on the first action component and the
+    other input data (observations).
+
+    Note, this RLModule works in combination with any algorithm, whose Learners require
+    the `ValueFunctionAPI`.
     """
 
     @override(RLModule)
-    @OverrideToImplementCustomLogic_CallToSuperRecommended
     def setup(self):
         super().setup()
 
-        # Build the encoder.
-        self.encoder = self.catalog.build_encoder(framework=self.framework)
-
-        # Build the prior and posterior heads.
-        # Note further, we neet to know the required input dimensions for
-        # the partial distributions.
-        self.required_output_dims = self.action_dist_cls.required_input_dim(
-            space=self.action_space,
-            as_list=True,
-        )
-        action_dims = self.action_space[0].shape or (1,)
-        latent_dims = self.catalog.latent_dims
-        prior_config = MLPHeadConfig(
-            # Use the hidden dimension from the encoder output.
-            input_dims=latent_dims,
-            # Use configurations from the `model_config`.
-            hidden_layer_dims=self.model_config["head_fcnet_hiddens"],
-            hidden_layer_activation=self.model_config["head_fcnet_activation"],
-            output_layer_dim=self.required_output_dims[0],
-            output_layer_activation="linear",
+        # Assert the action space is correct.
+        assert isinstance(self.action_space, gym.spaces.Tuple)
+        assert isinstance(self.action_space[0], gym.spaces.Discrete)
+        assert self.action_space[0].n == 3
+        assert isinstance(self.action_space[1], gym.spaces.Box)
+
+        self._prior_net = nn.Sequential(
```

Comment:
This new definition of the autoregressive module simplifies the setup and is better readable, I admit. However, it completely blanks out how users could/should use the `RLModule` attributes as well as the `Catalog` that makes setting up networks simply by pre-configured hyparparameters. Given that most current examples define the RLModule directly without leveraging its specific attributes, methods, or the `Catalog` - and recognizing that this often presents a challenge - we should evaluate our approach. Specifically, we should consider whether to (a) provide these same examples also in a form that explicitly demonstrate the use of these class attributes, methods, and the catalogue, or (b) streamline the `RLModule` by simplifying it further and potentially removing the `Catalog` entirely.

Reply:
By adding too many APIs to this example (Catalog) that have nothing to do with the actual problem to be solved here (autoregressive sampling of the 2 action components), we are cluttering this example.

I agree, we need to think about the usefulnes of `Catalog` in general. My take has been for a while now to get rid of it and simply replace it with utility functions that the user may use, but doesn't have to. Most users don't have the problem of 1 Model -> 100 spaces (like RLlib does!), but have a relatively fixed space and thus also a relatively stable model encoder architecture that doesn't require a complex "catalog decision tree".